rio/transmat/impl/git2/git_internals.go

```diff
@@ -0,0 +1,380 @@
+}
+
+/*
+	We force https when talking to github because github may refuse to respond to http urls.
+	Otherwise this will return an endpoint that will use the correct protocol based on the remote.
+	We could improve behavior by overriding installing new protocols.
+	Perhaps using a particular git user agent would avoid the strange github behavior.
+*/
+func gitCreateEndpoint(remote string) transport.Endpoint {
+	endpoint, err := transport.NewEndpoint(remote)
+	if err != nil {
+		panic(meep.Meep(
+			&rio.ErrInternal{Msg: "Failed to initiate git endpoint"},
+			meep.Cause(err),
+		))
+	}
+	if endpoint.Protocol() == "http" {
+		parsedUrl, err := url.Parse(remote)
+		if err != nil {
+			panic(meep.Meep(
+				&rio.ErrInternal{Msg: "Failed to parse git remote url"},
+				meep.Cause(err),
+			))
+		}
+		// Force https urls on github.com because github is silly and will not send back a response
+		if HasFoldedSuffix(parsedUrl.Hostname(), githubHostname) {
+			parsedUrl.Scheme = "https"
+		}
+		endpoint, err = transport.NewEndpoint(parsedUrl.String())
+		if err != nil {
+			panic(meep.Meep(
+				&rio.ErrInternal{Msg: "Failed to initiate git endpoint"},
+				meep.Cause(err),
+			))
+		}
+	}
+	return endpoint
+}
+
+/*
+	Executes an upload-pack-request on the remote
+	Loads the response into the provided storage
+*/
+func gitFetch(remote string, gitStore storer.Storer, uploadRequest *packp.UploadPackRequest) {
+	endpoint := gitCreateEndpoint(remote)
+	gitClient, err := client.NewClient(endpoint)
+	if err != nil {
+		panic(meep.Meep(
+			&rio.ErrInternal{Msg: "Failed to create git client"},
+			meep.Cause(err),
+		))
+	}
+	session, err := gitClient.NewUploadPackSession(endpoint, nil)
+	if err != nil {
+		panic(meep.Meep(
+			&rio.ErrInternal{Msg: "Failed to create git upload-pack session"},
+			meep.Cause(err),
+		))
+	}
+	response, err := session.UploadPack(uploadRequest)
+	if err != nil {
+		panic(meep.Meep(
+			&rio.ErrInternal{Msg: "git upload-pack-request failed"},
+			meep.Cause(err),
+		))
+	}
+	err = packfile.UpdateObjectStorage(gitStore, response)
+	if err != nil {
+		panic(meep.Meep(
+			&rio.ErrInternal{Msg: "Failed to update git object storage"},
+			meep.Cause(err),
+		))
+	}
+	// This operation may block if the response has not been processed.
+	err = session.Close()
+	if err != nil {
+		panic(meep.Meep(
+			&rio.ErrInternal{Msg: "Failed to close upload-pack-session"},
+			meep.Cause(err),
+		))
+	}
+}
+
+// Copied from go-git/worktree with minor changes
+func checkoutFile(f *object.File, fs billy.Filesystem) (err error) {
+	mode, err := f.Mode.ToOSFileMode()
+	if err != nil {
+		return
+	}
+
+	if mode&os.ModeSymlink != 0 {
+		return checkoutFileSymlink(f, fs)
+	}
+
+	from, err := f.Reader()
+	if err != nil {
+		return
+	}
+
+	defer ioutil.CheckClose(from, &err)
+
+	to, err := fs.OpenFile(f.Name, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, mode.Perm())
+	if err != nil {
+		return
+	}
+
+	defer ioutil.CheckClose(to, &err)
+
+	_, err = io.Copy(to, from)
```

Comment:
There's a decent example of this in the new Rio portages: https://github.com/polydawn/rio/blob/16835742dcbb49e44f786f3f26d5943e99961e68/transmat/tar/tar_packing.go#L91-L111

Major :key: : Make a method that makes a standardized Metadata object, stacks up what metadata git actually gives us on top of that, and then use the result when calling PlaceFile.

That example over yonder reuses a single Metadata object and mutates it to avoid allocations, and I think that's slightly preferable -- but we can ignore that here if you prefer since that's the callback-oriented form this is currently in makes the alloc avoidance near impossible (and doesn't matter terribly since presumably anywhere near here we're IO bound by a huge margin anyway).